### ISSUE TITLE
Python Language Server uses too much CPU

### ISSUE THREAD

**Issue by brettcannon** [Microsoft/python-language-server#479]
@VelizarVESSELINOV commented on [Tue Dec 04 2018](https://github.com/Microsoft/vscode-python/issues/3557)

## Environment data

- VS Code version: 1.29.1
- Extension version (available under the Extensions sidebar): 2018.11.0 
- OS and version: macOS Mojave 10.14.1
- Python version (& distribution if applicable, e.g. Anaconda): Python 3, Python 2...
- Type of virtual environment used (N/A | venv | virtualenv | conda | ...): N/A 
- Relevant/affected Python packages and their versions: N/A

## Expected behaviour

< 1% CPU

## Actual behaviour

\> 99% CPU

<img width="719" alt="image" src="https://user-images.githubusercontent.com/8209940/49480297-c936ff00-f7eb-11e8-9e17-3059cf8939fa.png">

<img width="636" alt="image" src="https://user-images.githubusercontent.com/8209940/49480306-d653ee00-f7eb-11e8-9d07-c9a08e066e67.png">

<img width="656" alt="image" src="https://user-images.githubusercontent.com/8209940/49480316-e075ec80-f7eb-11e8-8f55-162733fddf92.png">

<img width="999" alt="image" src="https://user-images.githubusercontent.com/8209940/49480387-203cd400-f7ec-11e8-9e4e-666e08145c75.png">

<img width="600" alt="image" src="https://user-images.githubusercontent.com/8209940/49480336-f1266280-f7eb-11e8-94f2-e8567a3896d3.png">

## Steps to reproduce:
1. Install several versions of Python
2. Run VS Code

## Logs
Output for `Python` in the `Output` panel (`View`→`Output`, change the drop-down the upper-right of the `Output` panel to `Python`)

```
Starting Microsoft Python language server.
Microsoft Python Language Server version 0.1.65.0
Initializing for /usr/local/bin/python3
```

---

@brettcannon commented on [Wed Dec 05 2018](https://github.com/Microsoft/vscode-python/issues/3557#issuecomment-444692454)

Do you have `"python.autoComplete.extraPaths"` defined?

---

@PushUpek commented on [Thu Dec 06 2018](https://github.com/Microsoft/vscode-python/issues/3557#issuecomment-444816311)

I have the same issue, but only with python2

---

@VelizarVESSELINOV commented on [Thu Dec 06 2018](https://github.com/Microsoft/vscode-python/issues/3557#issuecomment-445040708)

@brettcannon only in the default user settings defined as `[]`

---

@VelizarVESSELINOV commented on [Thu Dec 06 2018](https://github.com/Microsoft/vscode-python/issues/3557#issuecomment-445042898)

It is the biggest CPU consumer of all the processes: >21 hours

![image](https://user-images.githubusercontent.com/8209940/49613818-2490fa80-f96e-11e8-8c01-6ca9771985f7.png)

Side comment that maybe it is the case of @PushUpek I have a lot of files and folders in my current folder that is not Python code or libraries, questioning if Microsoft Python Server is not trying to parse GiB of data continuously in the current folder of VS Code project.

---

@jonaswitt commented on [Mon Dec 10 2018](https://github.com/Microsoft/vscode-python/issues/3557#issuecomment-445815569)

+1 - I have the same problem with 

VS Code 1.29.1
Microsoft Python Language Server version 0.1.65.0
Python Extension 2018.11.0
Python 2.7.15 (venv-based environment)

My `python.autoComplete.extraPaths` setting is also `[]`. I tried to disable Jedi and Linting. I keep killing the process to keep my Macbook's fans from spinning, but it always respawns at 100% CPU usage. 

Is there any way to disable the language server? I believe my only option right now is to disable the python extension completely.

---

@dwt commented on [Mon Dec 10 2018](https://github.com/Microsoft/vscode-python/issues/3557#issuecomment-445885664)

This happens for me too. Python 2.7 with a moderately big project. What Info can I provide to help debug this?

**Comment by narfanar** [comment]
On #3653 (duplicate)

Just in case the differences matter:

I'll have to check on CPU use, but I didn't notice it. I'd guess it's usually reasonable, even when memory use 10GB or higher (most of the time).

I'm on Python 3 not 2, and have only that version of Python installed.

**Comment by PushUpek** [comment]
I have couple projects. On python3 is ok. But on python2 cpu usage skips to 100%.

Language Server: 0.1.65

Python output, Extension Output, etc. Is empty, no error messages

**Comment by MikhailArkhipov** [comment]
Most probably similar to #495

**Comment by Spongman** [comment]
i can repro this:
- vscode 1.30.0
- ms-ms-python.python 4.5 (only extension enabled)
- `git https://github.com/mcedit/mcedit2.git --depth 1`
- hit `f12` on a symbol

**Comment by drorata** [comment]
Seems like I'm running into the same issue with scikit-learn. 100% and minimal RAM.

**Comment by MikhailArkhipov** [comment]
To all folks: we made some tactical fixes, but there is inherent issue to the analysis engine approach.

We are currently working on major restructuring of the static analysis engine that should address the issue with both CPU and memory. If you install Zenhub extension to Chrome, you should be able to track epic https://github.com/Microsoft/python-language-server/issues/425. The work is not trivial and is not a bug fix level change. 

WIP PR https://github.com/Microsoft/python-language-server/pull/465. 

Thanks you for using the LS, the feedback is much appreciated.

**Comment by Yajo** [comment]
Just as a simple not-so-workaround, how about assigning a lower priority to the spawned process?

On Linux all processes have a value called `nice` which indicates its priority. A "normal" user process usually uses nice=0. System important processes use lower nice, such as -20; you need to be root to lower the nice. Background, not-so-important processes are expected to use a higher value, such as nice=5.

By doing this very simple patch, we don't fix the CPU leak, but at least we tell the CPU to not block more important processes just for the sake of keep running the language server. Basically, the users wouldn't notice so much this leak. And in any case, it's a good practice for background processes.

I usually [`renice`](http://www.man7.org/linux/man-pages/man1/renice.1.html) the LS process and at least it doesn't disturb so much.

I bet other OS have a similar system.

**Comment by jakebailey** [comment]
That might be possible. Given this is in C#, we can do:

```csharp
Process.GetCurrentProcess().PriorityClass = ProcessPriorityClass.BelowNormal;
```

Which has the effect of setting the nice to 10 on Unix-y platforms.

**Comment by jakebailey** [comment]
On second thought, it seems like that'd only help in cases where the system is already using all of its cores at once. If the system has a free core to schedule the process on, it's going to do that even if the nice value is high. I'm not sure that would affect other processes if there are already free cycles...

**Comment by Spongman** [comment]
1) it can't hurt
2) cache-locality matters
3) "using all of its cores" happens more often than you might think on a real machine.

**Comment by musayev-io** [comment]
OS: macOS 10.14.2
VSCode: 1.30.2
MS-Python Extension: 2019.1.0
Microsoft Python Language Server version 0.1.78.0

I'm having a similar issue on my MBP, where Microsoft.Python.LanguageServer is running up to 100%+ CPU usage when I have large Python projects open (one, in particular, is ~200MB) within a virtual environment.

I have tested this with both Python 3.6.7 (from the official website), Python 2.7.15 (installed from Brew and another that's native to Mac). All versions in the virtual environment exhibit similar behavior. 

I have this `python.jediEnabled` set to `true` & `false`.
Additionally, I have linting disabled, so there's no additional overhead from that POV. 

Very frustrating, as my CPU is normal when I use Atom or PyCharm with the same project. Please let me know if I can help in any way to try and get to the bottom of this.

**Comment by jakebailey** [comment]
Please see this comment above: https://github.com/Microsoft/python-language-server/issues/479#issuecomment-453196291

The story hasn't changed; it's still being worked on. See here for specific issues: https://github.com/Microsoft/python-language-server/issues?q=label%3AGA

(I also responded in the other thread you commented on).

**Comment by MikhailArkhipov** [comment]
https://github.com/Microsoft/python-language-server/pull/546

**Comment by j9ac9k** [comment]
Please forgive me for commenting on a closed issue.

Congratulations on that pull request, that looks like a ton of work went into this, thank you.  I am curious if you know know what version of the Language Server will incorporate that pull request or when it will go live?

Thanks again for this awesome tool.

**Comment by MikhailArkhipov** [comment]
If you install Zenhub extension to Chrome, you should be able to see our schedule. Specifically, the Release report. 

https://github.com/Microsoft/python-language-server#workspaces/python-language-server-5b897936af3c0a2186966c0f/reports?report=release&release=5c469ca115d68f55b9388c22

**Comment by BilalDendani** [comment]
I am also have the same problem with the following version 
Version: 1.32.3
Commit: a3db5be9b5c6ba46bb7555ec5d60178ecc2eaae4
Date: 2019-03-14T23:38:49.842Z
Electron: 3.1.6
Chrome: 66.0.3359.181
Node.js: 10.2.0
V8: 6.6.346.32
OS: Linux x64 4.18.19-100.fc27.x86_64

**Comment by NiNJAD3vel0per** [comment]
this sucks !!!

**Comment by VelizarVESSELINOV** [comment]
@MikhailArkhipov @brettcannon 

It is exceptional to see so badly managing resource piece of software in production with known bugs (and/or bad design) since months and still not fixed. Not sure what it is wrong the management of the project or people working on this project or the testing, but it is very difficult for users of VS Code (Python on Mac...) to trust the capacity of Microsoft to produce good quality software.

What is the required hardware capacity to run VS Code with Python (on Mac)? 70 GiB RAM for VS Code?

![image](https://user-images.githubusercontent.com/8209940/57476616-97bfd500-725c-11e9-8fb2-0e9af69d64cc.png)

**Comment by jakebailey** [comment]
This is not normal, it is a bug. We have been working for many weeks on performance improvements, including memory usage. At the latest revision, analyzing tensorflow on my machine only uses 400 MB, as opposed to the multiple gigabytes it used to take. If you have a project which reproduces it, we'd love to test it and figure out why that would occur.

Also, try switching up to the beta release, which has many more improvements not yet in stable.

```json
"python.analysis.downloadChannel": "beta"
```

**Comment by jakebailey** [comment]
#832

**Comment by VelizarVESSELINOV** [comment]
@jakebailey I understand that this is a bug, I'm happy to see that you are working on fixing it, but the problem of abusive usage of resources is not new. This ticket is more six months old, and there was other as well earlier Q4 last year. I'm surprised that it is taking you more than six months to fix it, plus have the feeling that is going backward not reducing the resources used but increasingly. 20-40 GB was the "normal" level a few months ago, and now it is 70 GB.

I hope that soon you will fix it, and not again add small patches that solve few loading pick issues. The hidden requirement to restart many times per day VS Code reminds me of the old times of  Windows.

**Comment by esbozos** [comment]
![Captura de Pantalla 2019-05-12 a la(s) 2 27 46 a  m](https://user-images.githubusercontent.com/4162921/57579178-09b33c80-745e-11e9-93c1-05eb84a31044.png)

Analyzing workspace of Microsoft.Python.LanguageServer at the early start of opening, is a killing resources machines, is just like a python surrounding your neck, iI disabled other extensions, Jedi False, and the server just takes all available resources as a desperately hungry that just return from the jungle ...

Please let me know if you have some tips, tricks or guidance to resolve or how we can help to resolve this.

**Comment by esbozos** [comment]
Reading related reports issues, and testing, I got an improvement that helps me a few. 
Setting the download channel to "beta" version, the log level "Trace" and Jedi False btw. 

  "python.analysis.logLevel": "Trace",
   "python.analysis.downloadChannel": "beta",
   "python.jediEnabled": false

**Comment by ipstone** [comment]
Thank you for the work on python language server. 

I am a happy user of vscode, but when it comes to python code in my project, I have to switch to sublime etc., because python package is bundled together with language server (I might be wrong), and now I also have this issue of very high cpu usage, (I have relative faster macbook air with i7 processor)... 

Is it possible to separate the python extension and python language server?

**Comment by jakebailey** [comment]
You can disable the language server by setting `python.jediEnabled` to true. The language server is not enabled by default in the Python extension; you would have needed to enable it explicitly to have it run.

**Comment by chienjchienj** [comment]
Setting `python.jediEnabled` to true works for me

but it keeps showing this....
![image](https://user-images.githubusercontent.com/605097/60937914-864e8400-a304-11e9-9a2e-493d44ede64c.png)

Is there a way to ignore this permanently?

**Comment by jakebailey** [comment]
You can't use Jedi with IntelliCode; IntelliCode's Python support is an extension to the language server. That gear button may offer a way, but past that I do not know other than to remove IntelliCode.

Be aware that we've just built 0.3.22, which is available in the beta/daily channels. It contains some fixes which can help. You can set this and reload to update:

```
"python.analysis.downloadChannel": "beta"
```
